### PR TITLE
Adds Functionality for controlling ui with serial and hardware inputs

### LINKED ISSUE #34: models, controllers and state transition model
Now that we have the V in MVC, it's time to start thinking about how to hook up inputs and data to the graphical ui elements.  Key ideas to explore are:
- how to hook up inputs in such a way that we can easily reassign their functionality between different states of the dyno controller
- how to transition between states--hook up inputs, instantiate new resources, dispose of old ones
- how to push ui events to UIEventHandler

### PR DESCRIPTION
**Pull request by cosparks**
#### HardwareInput
- abstract class which provides a means of linking an input (button, encoder, serial, etc..) with a callback function
- classes which implement HardwareInput will check their respective i/o every control loop and call their input callback function when input is detected
- ```input_data_t``` (which is ```int32_t```) represents the input / change in input and is passed as argument to the input callback

#### InputManager
- provides interface for assigning input callbacks to inputs
- because input behavior will change between different states/menus, we will frequently need to reassign callbacks associated with each input
- Inputs are also identified by unique id in settings.h
 
#### ControllerBase
- base class for all ui controller types
- provides implementations of all ui input handler methods, which do nothing by default (this means that ui controllers will only have to implement input handlers for the inputs which that use -- methods which aren't implemented will safely do nothing)
- expose inFocus getter, which returns index of ui element which is currently focussed

#### ControllerMenu
- default menu controller class which will define most of the functionality we need for most simple UI menus

#### ControllerFactory
- deals with nasty circular dependencies between ApplicationContext and children of ControllerBase class
- handles construction of different ui Controllers, also currently contains button configurations for different MenuController

#### Application Context
- class which handles changes in context between different menus
- named after 'context' from [state pattern](https://en.wikipedia.org/wiki/State_pattern)

### Extras
- InputSerial class adds means of controlling menu focus/selection with up/down arrow keys and return key: open serial monitor and press keys to experiment
- Converts UIEventHandler to singleton class

### CODE REVIEW

**Comment by cosparks** [comment]
closes #34

File: src/System/HardwareInput.h [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        virtual ~HardwareInput() { }
+        virtual void init() = 0;
+        virtual void run() = 0;
+        void setAction(InputCallback action) { _action = action; }
+
+    protected:
+        InputCallback _action = [](int32_t) { }; // default action
+};
+
+class HardwareDemuxer {
```
Comment (LaPommeCosmique) [inline src/System/HardwareInput.h]: would you be able to add a comment explaining what this class does? Its not immediately clear to me

Comment (cosparks) [reply src/System/HardwareInput.h]: For sure.  Have you worked with multiplexers before?  They allow you to split up signals from one to many.  We were running low on pins for all the different functions of dyno, so i added multiplexer for all the ui button inputs which allows us to read 4 buttons while only using 3 pins

Comment (cosparks) [reply src/System/HardwareInput.h]: I actually wound up removing that class from this PR.  I've got another PR ready to go for when this one is done which concerns reading the demuxxed inputs

**Review by LaPommeCosmique — APPROVED** [review]

File: src/System/HardwareInput.h (hunk 1 above)
Comment (LaPommeCosmique) [reply src/System/HardwareInput.h]: Ahh okay. So the multiplexer would be a hardware component that is represented by this HardwareDemuxer class?

Comment (cosparks) [reply src/System/HardwareInput.h]: exactly

**Comment by cosparks** [comment]
@LaPommeCosmique the scope of this PR has definitely ballooned beyond what I originally intended.  I've updated the notes to include the more recent changes I've made.

Sorry about that

**Comment by cosparks** [comment]
I'm going to have another look tonight to make sure I didn't do anything stupid this morning, then I'll merge and close

**Pull request merged**